theano/sandbox/cuda/cuda_ndarray.cu

```diff
@@ -4393,15 +4393,42 @@ int CudaNdarray_sgemv(float alpha, const CudaNdarray * A, const CudaNdarray * B,
     int sb_0 = (CudaNdarray_HOST_DIMS(B)[0] > 1) ? CudaNdarray_HOST_STRIDES(B)[0] : 1;
     int sc_0 = (CudaNdarray_HOST_DIMS(C)[0] > 1) ? CudaNdarray_HOST_STRIDES(C)[0] : 1;
 
-    if (sa_0 == 0)
-        sa_0 = 1;
-    if (sa_1 == 0)
-        sa_1 = 1;
+    if (sa_0 == 0) sa_0 = 1;
+    if (sa_1 == 0) sa_1 = 1;
 
     // This is important because we can end up not calling Sgemv at all
     cublasStatus_t err = CUBLAS_STATUS_SUCCESS;
     if (CudaNdarray_SIZE(C)) {
-        if ((CudaNdarray_HOST_DIMS(A)[0] <= 1)
+        // A is row vector & alpha==1 & beta==0 -> use cublasSdot
+        if (CudaNdarray_HOST_DIMS(A)[0] == 1 && alpha==1.f && beta==0.f) {
+            //TODO: this is a rather temporary solution
+            // 1. better temp solution:
+            // replace this with custom inner product kernel with
+            // alpha and beta parameter
+            // 2. permanant solution:
+            // define a new "InnerProduct" Op, add an optimization
+            // "gemv -> inner_prod", perhaps for CPU/GPU both
+            float* dev_dst = CudaNdarray_DEV_DATA(C)+1-sc_0;
+            cublasPointerMode_t pmode;
+            cublasGetPointerMode(handle, &pmode);
+            // need to store dot result on device here
+            cublasSetPointerMode(handle, CUBLAS_POINTER_MODE_DEVICE);
```

Comment:
If I don't add this, this will cause CUBLAS take dot result as a host pointer, causing segfault ([Example](http://stackoverflow.com/questions/28979616/segmentation-fault-while-executing-cuda-program)). I set it back to make sure other CUDA code still function correctly.